specification/under_development/profiles/Adaptive_Profile.md

```diff
@@ -0,0 +1,1028 @@
+profile based on optional features, but at minimum, the simple `br` branch
+instruction must be supported.
+
+### Bullet 5: Mid-circuit measurement
+
+The adaptive profile allows for mid-circuit measurement to be expressed in
+programs by removing restrictions on a measurement operation (we will use
+measurement in Z represented by `llvm __quantum__qis__mz__body` in this
+document, but other forms of measurement can also apply) that state that
+measurements occur at the end of the quantum program and that the only
+instructions following them are output recording functions. If a backend opts
+into using this feature then it allows for programs with the following structure:
+
+```llvm
+call quantum__qis__h__body(%Qubit* null) ; arbitrary gate here
+...
+call quantum__qis__mz__body(%Qubit* null, %Result* null)
+call quantum__qis__h__body(%Qubit* nonnull inttoptr (i64 1 to %Qubit*)) ; arbitrary gate choice again
+...
+```
+
+### Bullet 6: Conditional Quantum Operations
+
+Although forward branching can be useful when combined with purely classical
+operations within a quantum program, the real utility is being able to
+conditionally perform gates as part of things like repeat-until-success or
+quantum-error-correction routines in quantum programs.
+
+Here is an example program that mixes mid-circuit measurement (**Bullet 5**), a
+`qis` function to convert a measurement result to a boolean (**Bullet 7**), and
+a branch instruction (**Bullet 4**) to control how gates are applied based on
+mid-circuit measurement.
+
+```llvm
+  tail call void @__quantum__qis__h__body(%Qubit* null)
+  tail call void @__quantum__qis__mz__body(%Qubit* null, %Result* null)
+  tail call void @__quantum__qis__reset__body(%Qubit* null)
+  %0 = tail call i1 @__quantum__rt__read_result__body(%Result* null)
+  br i1 %0, label %then, label %continue
+then:                                   ; preds = %entry
+  tail call void @__quantum__qis__x__body(%Qubit* nonnull inttoptr (i64 1 to %Qubit*))
+  br label %continue__1.i.i.i
+
+continue:
+...
+```
+
+### Bullet 7: Measurement Results to Booleans
+
+A function to turn measurement results into classical data, mid-circuit, is
+critical to using forward branching and conditional quantum execution.
+
+```llvm
+%0 = tail call i1 @__quantum__rt__read_result__body(%Result* null)
+  br i1 %0, label %then, label %continue
+```
+
+## Optional Capabilities
+
+### Bullet 8: Qubit resetting
+
+If a backend opts into supporting this feature, then we can allow
+for an instruction to reset qubits back to the |0> state to be defined
+in the instruction set. This can be illustrated by the following program:
+
+```llvm
+call quantum__qis__h__body(%Qubit* null) ; arbitrary gate here
+...
+call quantum__qis__reset__body(%Qubit* null)
+call quantum__qis__h__body(%Qubit* nonnull inttoptr (i64 1 to %Qubit*)) ; arbitrary gate choice again
+...
+```
+
+This can be combined with the mid-circuit measurement (**Bullet 5**) to allow
```

Comment:
I don't get the argument that reset is necessary to allow for qubit reuse. Reset is just an optimized implementation of
```
call quantum__qis__mz__body(%Qubit* null, %Result* null)
%0 = tail call i1 @__quantum__rt__read_result__body(%Result* null)
br i1 %0, label %then, label %continue
then:  
  tail call void @__quantum__qis__x__body(%Qubit* null)
  br label %continue
continue:
...
```